src/host/services/IMUService.cpp

```diff
@@ -26,38 +26,204 @@
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  switch (_config.mounting) {
+    case VerticalPortHull:
+      _roll = SKDegToRad(eulerAngles.z());
+      _pitch = SKDegToRad(eulerAngles.y());
+      _heading = SKDegToRad(fmod(eulerAngles.x() + 270, 360));
+    break;
+    case VerticalStbHull:
+      _roll = SKDegToRad(eulerAngles.z())*(-1);
+      _pitch = SKDegToRad(eulerAngles.y())*(-1);
+      _heading = SKDegToRad(eulerAngles.x());
+    break;
+    case VerticalTopToBow:
+      _roll = SKDegToRad(eulerAngles.y());
+      _pitch = SKDegToRad(eulerAngles.z());
+      _heading = SKDegToRad(fmod(eulerAngles.x() + 180, 360));
+    break;
+    case HorizontalLeftSideToBow:
+      _roll = SKDegToRad(eulerAngles.y());
+      _pitch = SKDegToRad(eulerAngles.z()) * (-1);
+      _heading = SKDegToRad(eulerAngles.x());
+    break;
+  }
+
+  // update NavigationHeadingMagnetic if quality (= calibration value) is ok
+  if (isMagCalibrated()) {
+    update.setNavigationHeadingMagnetic(_heading);
+  }
+  // update NavigationAttitude if quality (= calibration value) is ok
+  if (isHeelAndPitchCalibrated()) {
+    update.setNavigationAttitude(SKTypeAttitude(/* roll */ _roll + _offsetRoll, /* pitch */ _pitch + _offsetPitch, /* yaw */ SKDoubleNAN));
+    //DEBUG("Heel = %.3f | Pitch = %.3f | Heading = %.3f", SKRadToDeg( _roll + _offsetRoll), SKRadToDeg( _pitch + _offsetPitch), SKRadToDeg(_heading));
+  }
+  _skHub.publish(update);
+
+  // Save calibration values to EEPROM every 30 Minutes, if BNO055 is fully calibrated
+  if (bno055.isFullyCalibrated() && _timeSinceLastCalSave > 1800000) {
+    if (saveCalibrationOffsets())
+      _timeSinceLastCalSave = 0;
+  }
+}
+
+// get the actual values of calibration, heel & pitch (including offset) and heading for display
+// all angles in radians
+void IMUService::getLastValues(int &sysCalibration, int &accelCalibration, double &pitch, double &roll, int &magCalibration, double &heading){
+  sysCalibration = _sysCalib;
+  accelCalibration = _accelCalib;
+  pitch = _pitch + _offsetPitch;
+  roll = _roll + _offsetRoll;
+  magCalibration = _magCalib;
+  heading = _heading;
+}
+
+//  Calc offset for making Heel=0 and Pitch=0
+//  (e.g. called with long button press in IMUMonitorPage)
+//  If an setOffset will be done a second time within 5 Seconds,
+//  the values will be stored into EEPROM and loaded at start of KBox
+void IMUService::setHeelPitchOffset() {
+  bool changed = false;
+  if (abs(_roll) < M_PI / 2 ) {
+    _offsetRoll = _roll * (-1);
+    changed = true;
+  }
+  if (abs(_pitch) < M_PI / 2 ) {
+    _offsetPitch = _pitch * (-1);
+    changed = true;
+  }
+
+  if ( changed ) {
+    INFO("Offset changed: Heel -> %.3f | Pitch -> %.3f", SKRadToDeg(_offsetRoll), SKRadToDeg(_offsetPitch));
+    if (_timeSinceLastSetOffset < 5000) {
```

Comment:
I do not understand why you have `<5000` here. That means you only save if it's been less than 5sec that we started? That means you cannot save if you wait more than 5s after booting?

Reply:
The first idea was, that with the first "long press" the offset is made temporary and a second "long press" will store it to flash.
Now I think it makes no sense, so the offset could be stored at first press.
But, I think I like the idea, that a second "long press" could reset offset to zero.